distribution/lib/Standard/Base/0.0.0-dev/src/Data/Enso_Cloud/Enso_File.enso

```diff
@@ -0,0 +1,196 @@
+        _ = [open_options, action]
+        Unimplemented.throw "Writing to Enso_Files is not currently implemented."
+
+    ## PRIVATE
+       ADVANCED
+       Creates a new input stream for this file and runs the specified action
+       on it.
+
+       Arguments:
+       - open_options: A vector of `File_Access` objects determining how to open
+         the stream. These options set the access properties of the stream.
+       - action: A function that operates on the input stream and returns some
+         value. The value is returned from this method.
+
+       The created stream is automatically closed when `action` returns (even
+       if it returns exceptionally).
+    with_input_stream : Vector File_Access -> (Input_Stream -> Any ! File_Error) -> Any ! File_Error | Illegal_Argument
+    with_input_stream self open_options action =
+        if self.is_directory then Error.throw (Illegal_Argument.Error "Enso_File folders cannot be opened as a stream." self.uri) else
+            if (open_options !=  [File_Access.Read]) then Error.throw (Illegal_Argument.Error "Enso_File files can only be opened for reading." self.uri) else
+                extension = if self.name.contains "." then "." + self.name.take (Text_Sub_Range.After_Last ".") else ""
+                url = "https://production-enso-organizations-files.s3.eu-west-1.amazonaws.com/" + self.organisation + "/" + self.id + extension
+                response = HTTP.fetch url HTTP_Method.Get []
+                response.if_not_error <| response.with_stream action
+
+    ## ALIAS load, open
+       GROUP Input
+       Read a file using the specified file format
+
+       Arguments:
+       - format: A `File_Format` object used to read file into memory.
+         If `Auto_Detect` is specified; the provided file determines the specific
+         type and configures it appropriately. If there is no matching type then
+         a `File_Error.Unsupported_Type` error is returned.
+       - on_problems: Specifies the behavior when a problem occurs during the
+         function.
+         By default, a warning is issued, but the operation proceeds.
+         If set to `Report_Error`, the operation fails with a dataflow error.
+         If set to `Ignore`, the operation proceeds without errors or warnings.
+    @format File_Format.default_widget
+    read : File_Format -> Problem_Behavior -> Any ! Illegal_Argument | File_Error
+    read self format=Auto_Detect (on_problems=Problem_Behavior.Report_Warning) =
+        _ = on_problems
+        case format of
+            Auto_Detect -> case self.asset_type of
+                Enso_Asset_Type.Folder -> self.list
+                Enso_Asset_Type.Project -> Error.throw (Illegal_Argument.Error "Projects cannot be read within Enso code. Open using the IDE.")
+                Enso_Asset_Type.Connection -> Error.throw (Illegal_Argument.Error "Connections cannot be read within Enso code. Open using the IDE.")
```

Comment:
What does it mean to "Open" a Connection using the IDE?

Reply:
In due course (not in this PR), a connection will define a link to a data source for example to an S3 file etc.
Opening a connection will read this in.

Reply:
I see but I still don't think I understand what does it mean to "Open" a connection?

Will it not be like a function e.g. `Connection.read`?

Reply:
When you `Data.read` a connection it will just read the target. 

So if it targets an S3 CSV file, the connection would be a JSON definition which tells us where the S3 file is. Opening it will get the JSON and then read the S3 file into Enso. Eventually, it will be via a proxy running in cloud but not in the initial versions.

Reply:
Ok, so this works +- how I thought it does.

In that context, I don't understand the `Connections cannot be read within Enso code. Open using the IDE.` message. How does reading the connection in Enso code differ from doing so in the IDE?

I understand that we cannot read a project in code, we want to _open_ the other project instead in that case - that is a difference instead.

But from perspective of the connection, I cannot figure out what is the difference between 'enso code' and 'opening in IDE' (which executes the _enso code_) and so I don't understand what this error message is indicating.

Reply:
Only that the connection is not implemented yet and I left that as a place-holder for now.

Project will probably be supported for reading as well but haven't done that yet. And yes error messages are still unclear.

Reply:
Yes - just working progress line for the moment but will amend.